programs/sunrise_beam/src/state.rs

```diff
@@ -1,65 +1,78 @@
+    /// The Sunrise Gsol mint.
     pub gsol_mint: Pubkey,
-    /// The Sunrise gsol mint authority bump
+    /// The starting Gsol supply.
+    pub pre_supply: u64,
+    /// Bump of the gsol mint authority PDA.
     pub gsol_mint_authority_bump: u8,
-    /// The beam controller yield account.
+    /// The Sunrise yield account.
     pub yield_account: Pubkey,
-    /// Holds [Allocation] details for beams.
-    pub allocations: Vec<Allocation>,
+    /// The factor to increase space by during a resize.
+    pub alloc_window: u8,
+    /// Holds [BeamDetails] for all supported beams.
+    pub allocations: Vec<BeamDetails>,
 }
 
-/// The allocation details for a beam.
-#[derive(AnchorSerialize, AnchorDeserialize, Clone, Default)]
-pub struct Allocation {
+/// Holds information about a registed beam.
+#[derive(AnchorSerialize, AnchorDeserialize, Clone, Default, Eq, Hash, PartialEq)]
+pub struct BeamDetails {
     /// Expected signer for mint and burn requests.
     pub beam: Pubkey,
     /// This beam's allocation expressed as a percentage.
     pub allocation: u8,
-    /// A beam in drain accepts withdrawals but no deposits.
+    /// The total amount of Gsol this beam has minted.
+    pub minted: u64,
+    /// A beam in drain accepts withdrawals but not deposits.
     pub draining_mode: bool,
 }
 
-impl Allocation {
-    pub const SPACE: usize = 32 + 1 + 1;
+impl BeamDetails {
+    /// Space in bytes of a borsh-serialized [BeamDetails] struct.
+    pub const SPACE: usize = 32 + 1 + 8 + 1;
 
     pub fn new(beam: Pubkey, allocation: u8) -> Self {
-        Allocation {
+        BeamDetails {
             beam,
             allocation,
+            minted: 0,
             draining_mode: false,
         }
     }
 }
 
-impl ControllerState {
-    pub fn size(initial_capacity: usize) -> usize {
-        8 + //discriminator
-        32 + 32 + 1 + 32 + (4 + (Allocation::SPACE * initial_capacity))
+impl State {
+    /// Calculate the space that would be allocated to hold a `beam_count`
+    /// number of beams.
+    pub fn size(beam_count: usize) -> usize {
+        8 /*discriminator*/
+        + 32 + 32 + 8 + 1 + 32 + 1 + (4 + (BeamDetails::SPACE * beam_count))
     }
 
+    /// Register a new [State] with the given information.
     pub fn register(
         &mut self,
         input: RegisterStateInput,
         gsol_mint_auth_bump: u8,
         gsol_mint: &Pubkey,
+        gsol_mint_supply: u64,
     ) {
         self.update_authority = input.update_authority;
         self.yield_account = input.yield_account;
         self.gsol_mint = *gsol_mint;
+        self.pre_supply = gsol_mint_supply;
         self.gsol_mint_authority_bump = gsol_mint_auth_bump;
-        self.allocations = Vec::with_capacity(input.initial_capacity);
-        self.allocations = Vec::default(); // Todo: Use Vec::with_capacity()?
+        self.alloc_window = input.alloc_window;
+        self.allocations = vec![BeamDetails::default(); input.initial_capacity];
```

Comment:
Will this fill the vector with empty instances of beamdetails? I think we probably rather want with_capacity, i.e. the size will be 0 but the capacity will be input.initial_capacity

Reply:
Deserialization would result in a vector of length 0 and capacity 0. I did it this way so we wouldn't need a field tracking the capacity. We just always make sure the vector is filled up to the max and know that a beam can be added if there's a BeamDetails::default() that it can replace

Reply:
ok - please document this then in comments wherever needed.

Reply:
Alright. Will do so